compiler/desugared/disambiguate.ml

```diff
@@ -0,0 +1,78 @@
+(* This file is part of the Catala compiler, a specification language for tax
+   and social benefits computation rules. Copyright (C) 2020 Inria, contributor:
+   Louis Gesbert <louis.gesbert@inria.fr>
+
+   Licensed under the Apache License, Version 2.0 (the "License"); you may not
+   use this file except in compliance with the License. You may obtain a copy of
+   the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+   License for the specific language governing permissions and limitations under
+   the License. *)
+
+open Shared_ast
+open Ast
+
+let expr ctx env e =
+  (* The typer takes care of disambiguating: this consists in: - ensuring
+     [EAbs.tys] doesn't contain any [TAny] - [EDStructAccess.name_opt] is always
+     [Some] *)
+  (* Intermediate unboxings are fine since the last [untype] will rebox in
+     depth *)
+  Typing.check_expr ctx ~env (Expr.unbox e)
```

Comment:
Where is the logic that throws an error if you're trying to type `x.f1` but `x` is of type struct `A` and `f1` belongs to struct `B`?

Reply:
You found it already ;) but precisely and for reference, here:
https://github.com/CatalaLang/catala/blob/4ee4a96ac7e9f12fe5cf1e7a4be8b78ad624d472/compiler/shared_ast/typing.ml#L467-L473

Ideally, we would suggest existing fields that are close syntactically, but that would be useful in many place, so could be a feature for the upcoming error message handling ? We could also list the structures that the specified field actually belongs to, if any :thinking:

Reply:
Actually, if the fields is uniquely resolved to a different struct, as in your example, we could trigger the "normal" type error message with the field requiring type `B` and the struct being type `A`. Do you think that would be more helpful in general ?

Reply:
Ideally the error message is customized saying something like "here you want to access a field of struct `B` but the value underlined below has the type of struct `A`"

Reply:
indeed that'll be nicer, I'll add a specific message for this case!